standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+## Docker standards
+
+### General
+- all containers should be created using Docker
+- Linux containers should be used
+- containers should not be run using root user
+- containers should be scanned regularly for vulnerabilities
+
+### Parent images
+We should create Dockerfiles that are extended from a minimal Defra created parent images.  This will allow us to benefit from improved security and more efficient builds as we will not have to repeat steps that are common to all Dockerfiles.
```

Comment:
There's an implication here that _someone_ is responsible for patching, maintaining, controlling and publishing these parent images.  I'm not sure that we're really there yet.

And I'm still not 100% convinced that having a pre-built image is the best way of achieving this.  I can see that it's useful for a project or programme to avoid having to keep creating the same base image all the time, but a department-wide base image feels like it could be a bit of a maintenance headache, as opposed to just having, say, a standard set of `Dockerfile`s that everyone can use to build their base images from, which would be easier to maintain.

Reply:
A fair comment. I was coming at it from a microservice architecture perspective where there'd end up being many similar Dockerfiles in multiple repos.  If there was a need to apply a security fix to counteract something in the base image the effort to change all of them would be significant.  Whereas using a base image it would be reduced.

Perhaps I should change this so this approach isn't mandatory but recommended for a microservice approach.

Not sure on your last comment, "as opposed to just having, say, a standard set of Dockerfiles that everyone can use to build their base images from, which would be easier to maintain." as that's basically what I was proposing.  The parent image would just be a Dockerfile with all the basic stuff setup best practice that others would use as a base to their own Dockerfiles.

Reply:
Yeah sorry, re-reading it I think my confusion stems from different uses of the terms **Dockerfile** and **image**.

So I would agree that it would be helpful to maintain a core set of `Dockerfile`s in GitHub that can be re-used by  projects, but I'm not convinced about having "minimal Defra created parent images" - I'm not 100% clear on which one this standard is suggesting, as it appears to be saying both.

Either way there's still a maintenance headache, but my feeling is that it's a lot less of a headache to just maintain some common `Dockerfile`s than it is to have a common container registry of cached images.  But by the sounds of it that is what you're proposing anyway?

Regarding the microservice architecture point, I think this probably needs clarifying. If a particular product is going to take a microservices approach, then that product needs a coordinated governance mechanism and that will likely include a shared container registry. However, that doesn't necessarily mean that we can have a common registry _between_ different products, so a common `Dockerfile` for Defra seems appropriate, with probably duplicated images built from that Dockerfile across different products.

Reply:
Yes that is what I'm proposing.  We'd have a GitHub repo which has our parent Dockerfiles.  Those files would have a CI/CD pipeline that pushes the image they create into a container registry.  Any service wanting to use our parent Dockerfile in their Dockerfile would specify that they are building from our parent image as opposed to say the Node image hosted in DockerHub.

eg

`FROM defra-container-registry/defra-node-parent`

The complication is where do we put that container registry, DockerHub is the easiest as all our services will be able to pull from it and the general community will benefit if they so wish.  Alternatively we could use ACR or ECR in Azure and AWS respectively but then we'll need to manage access across projects which may be problematic.

Reply:
I'll reword these docs to make it clearer when I mean images and when I mean files.

Reply:
> Simplest solution is as you say, a GitHub repo with Dockerfiles in may suffice but does make them more difficult to use and teams will probably end up just using their own base images in their own registry anyway.

Yes I think this is what I'm leaning towards formally advocating.  It's a bit of a trade-off between providing the most helpful thing possible (which I agree would be pre-built base images in a common registry) and not having one, single _thing_ that is a dependency for everything we do, which then inherits all of the varying needs of all of the different teams and impacts on all of them when it changes.

To me, it feels like a good idea to decouple the common container _definition_ from all the concerns about managing a registry of images and all the non-functionals that go with that.  Teams will be implementing/using their own container registries anyway, so we're probably not saving them a huge amount of effort, albeit it's a little bit easier to build FROM an existing image than to have your own project that builds it and puts it in your registry.

Reply:
Ok I can go with that.  It may be in time we naturally evolve towards a grander plan,  but that seems a good starting position.  

Side note,  we have some Node.js and .Net Core parent images in a public GitHub repo already.